pkg/agent/openflow/fields.go

```diff
@@ -0,0 +1,128 @@
+	CustomReasonDenyRegMark    = binding.NewRegMark(CustomReasonField, CustomReasonDeny)
+
+	// reg1(NXM_NX_REG1)
+	// Field to cache the ofPort of the OVS interface where to output packet.
+	TargetOFPortField = binding.NewRegField(1, 0, 31, "TargetOFPort")
+
+	// reg2(NXM_NX_REG2)
+	// Field to help swap values in two different flow fields in the OpenFlow actions. This field is only used in func
+	// `arpResponderStaticFlow`.
+	SwapField = binding.NewRegField(2, 0, 31, "SwapValue")
+
+	// reg3(NXM_NX_REG3)
+	// Field to store the selected Service Endpoint IP
+	EndpointIPField = binding.NewRegField(3, 0, 31, "EndpointIP")
+	// Field to store the conjunction ID which is for "deny" rule in CNP. It shares the same register with EndpointIPField,
+	// since the service selection will finish when a packet hitting NetworkPolicy related rules.
+	CNPDenyConjIDField = binding.NewRegField(3, 0, 31, "CNPDenyConjunctionID")
+
+	// reg4(NXM_NX_REG4)
+	// reg4[0..15]: Field to store the selected Service Endpoint port.
+	EndpointPortField = binding.NewRegField(4, 0, 15, "EndpointPort")
+	// reg4[16..18]: Field to store the state of a packet accessing a Service. Marks in this field include,
+	//	- 0b001: packet need to do service selection.
+	//	- 0b010: packet has done service selection.
+	//	- 0b011: packet has done service selection and the selection result needs to be cached.
+	ServiceEPStateField = binding.NewRegField(4, 16, 18, "EndpointState")
+	EpToSelectRegMark   = binding.NewRegMark(ServiceEPStateField, 0b001)
+	EpSelectedRegMark   = binding.NewRegMark(ServiceEPStateField, 0b010)
+	EpToLearnRegMark    = binding.NewRegMark(ServiceEPStateField, 0b011)
+	// reg4[0..18]: Field to store the union value of Endpoint port and Endpoint status. It is used as a single match
+	// when needed.
+	EpUnionField = binding.NewRegField(4, 0, 18, "EndpointUnion")
+
+	// reg5(NXM_NX_REG5)
+	// Field to cache the Egress conjunction ID hit by TraceFlow packet.
+	TFEgressConjIDField = binding.NewRegField(5, 0, 31, "TFEgressConjunctionID")
+
+	// reg(N6XM_NX_REG6)
+	// Field to store the Ingress conjunction ID hit by TraceFlow packet.
+	TFIngressConjIDField = binding.NewRegField(6, 0, 31, "TFIngressConjunctionID")
+)
+
+// Fields using xxreg.
+var (
+	// xxreg3(NXM_NX_XXREG3)
+	// xxreg3: Field to cache Endpoint IPv6 address. It occupies reg12-reg15 in the meanwhile.
+	EndpointIP6Field = binding.NewXXRegField(3, 0, 127)
+)
+
+// Marks using CT.
+var (
+	// Mark to indicate the connection is initiated through the host gateway interface
+	// (i.e. for which the first packet of the connection was received through the gateway).
+	FromGatewayCTMark = binding.NewCtMark(0x20, 0, 31)
+	// Mark to indicate DNAT is performed on the connection for Service.
+	ServiceCTMark = binding.NewCtMark(0x21, 0, 31)
+)
+
+// Fields using CT label.
+var (
+	// Field to store the ingress rule ID.
+	IngressRuleCTLabelField = binding.NewCtLabelField(0, 31, "ingressRuleCtLabel")
```

Comment:
You didn't name CT marks FromGatewayCTMarkField and relevant functions NewCtMarkField, why adding "Field" suffic to CT labels?

Reply:
In my thought, Mark and Field are different concepts. A field only requires a bit range, but no fixed value, and Antrea can use the Field to match/load any value. But for a Mark, it represents a fixed value and the value uses a field. For ctLabel, I didn't think it is used as a mark for fixed value, we only use it as a field for temp value in a connection.

Reply:
Besides, ct_mark is the field name used in OVS, like ct_label is a field name in OVS. It is weird to use CTMarkMark to represent a special mark using part bits of ct_mark But if you insist, I would change it.

Reply:
Got it. I misunderstood `FromGatewayCTMark`, but I see in the latest code you have removed `Field` from `CTLabel` struct, which may bring confusion and make it hard to extend `CTMark` in the future. When reg, ct_mark, ct_label are all fields of OVS, the fields and values are named in different ways:
| | reg | ct_mark | ct_label |
|  ----  | ----  | ---- | ---- |
| Field Struct | RegField | - | CTLabel |
| Value Struct | RegMark | CTMark | - |

It would be hard to understand that `XXXCTMark` is the struct of value while `XXXCTLabel` is the struct of field. And how we name the field struct of ct_mark in the future when we want to match a value on several bits like reg and ct_label? I think we cannot say ct_mark will only be used as a whole because it would be a waste of the limited connection level storage. In theory `FromGatewayCTMark` and `ServiceCTMark` should just use two bits, instead of 32 bits.

I think your original code makes more sense and we can name the struct of ct_mark related field to `CTMarkField` in the future, just like `CTLabelField`.

Reply:
Sure, but I would like to involve it in the future PRs.